### PR TITLE
0075 render floor plan

### PR DESCRIPTION
**Pull request by Steppschuh**
WIP for #75

### CODE REVIEW

File: app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMapBackground.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    }
+
+    public Point getPoint(Location location) {
+        return getPoint(location, topLeftLocation, topLeftPoint, metersPerPixel, bearing);
+    }
+
+    public static float getMetersPerPixel(Location firstReferenceLocation, Point firstReferencePoint, Location secondReferenceLocation, Point secondReferencePoint) {
+        double distanceInPixels = getPixelDistance(firstReferencePoint, secondReferencePoint);
+        double distanceInMeters = firstReferenceLocation.getDistanceTo(secondReferenceLocation);
+        return (float) (distanceInMeters / distanceInPixels);
```
Comment (marvinmirtschin) [inline app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMapBackground.java]: Can throw an `ArithmeticException` for division by 0 as `distanceInPixels` can be 0.

File: app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMapBackground.java [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    }
+
+    public static double getBearing(Location firstReferenceLocation, Point firstReferencePoint, Location secondReferenceLocation, Point secondReferencePoint) {
+        double locationBearing = getBearing(firstReferenceLocation, secondReferenceLocation);
+        double pointBearing = getBearing(firstReferencePoint, secondReferencePoint);
+        return ((locationBearing - pointBearing) + 360) % 360;
+    }
+
+    public static double getBearing(Point firstPoint, Point secondPoint) {
+        double angle = Math.atan2(secondPoint.y - firstPoint.y, secondPoint.x - firstPoint.x) * 180 / Math.PI;
```
Comment (marvinmirtschin) [inline app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMapBackground.java]: `Atan2` has a value range from `Pi` to `-Pi`. If we divide by `Pi` we will have a range of -1 to 1. Multiplying with 180 gives -180 to 180. 
If we now add 90 we will be in a range from -90 to 270. Modulo 360 will leave the value range in between -90 and 270.

File: app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMapBackground.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+    public static double getPixelDistance(Point firstPoint, Point secondPoint) {
+        int deltaX = Math.abs(secondPoint.x - firstPoint.x);
+        int deltaY = Math.abs(secondPoint.y - firstPoint.y);
+        return Math.sqrt((deltaX * deltaX) + (deltaY * deltaY));
+    }
+
+    public static Location getLocation(Point point, Location referenceLocation, Point referencePoint, double metersPerPixel, double backgroundBearing) {
+        double distance = metersPerPixel * getPixelDistance(referencePoint, point);
+        double bearing = (getBearing(referencePoint, point) + backgroundBearing + 360) % 360;
```
Comment (marvinmirtschin) [inline app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMapBackground.java]: As bearings can not be negative (at least they should the way we use them) the `+360` is unnecessary. (positive number + positive number will stay positive).

File: app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMapBackground.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    public static Point getPoint(Location location, Location referenceLocation, Point referencePoint, double metersPerPixel, double backgroundBearing) {
+        double distanceInMeters = location.getDistanceTo(referenceLocation);
+        double distanceInPixels = distanceInMeters / metersPerPixel;
+        double locationBearing = getBearing(referenceLocation, location);
+        double bearing = (locationBearing - backgroundBearing + 360) % 360;
+        return getShiftedPoint(referencePoint, distanceInPixels, bearing);
+    }
+
+    public static Point getShiftedPoint(Point referencePoint, double distanceInPixels, double bearing) {
+        double angleInRadians = Math.toRadians(bearing + 90);
```
Comment (marvinmirtschin) [inline app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMapBackground.java]: Where does the `+90` come from?

File: app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMap.java [hunk 5]
```diff
@@ -84,6 +87,26 @@ protected void onDraw(Canvas canvas) {
...
+    protected void drawBackground(Canvas canvas) {
+        super.drawBackground(canvas);
+
+        if (mapBackground == null) {
+            return;
+        }
+
+        Matrix matrix = new Matrix();
+        matrix.postRotate((float) mapBackground.getBearing());
+        matrix.postTranslate(-1700, -1350);
```
[the remark this answers is not in the document]
Comment (Steppschuh) [reply app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMap.java]: This is the work in progress part of this PR :)

**Comment by Steppschuh** [comment]
demo_app

Implementation works, but still has issues (probably related to padding in the `CanvasProjection`)

File: app/src/test/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMapBackgroundTest.java [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+        angle = 180;
+        shiftedPoint = BeaconMapBackground.getShiftedPoint(referencePoint, distance, angle);
+        assertPointEquals(new Point(100, 150), shiftedPoint, 0);
+
+        angle = 270;
+        shiftedPoint = BeaconMapBackground.getShiftedPoint(referencePoint, distance, angle);
+        assertPointEquals(new Point(50, 100), shiftedPoint, 0);
+    }
+
+    public static void assertPointEquals(Point expectedPoint, Point actualPoint, double delta) {
```
Comment (Uwinator) [inline app/src/test/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMapBackgroundTest.java]: Please re-arrange/move method to bottom. I would not make this public...

File: app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMapBackground.java (hunk 2 above)
Comment (Steppschuh) [reply app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMapBackground.java]: There's a test that asserts a value > 270:

```java
@Test
public void getBearing() {
    double bearing = BeaconMapBackground.getBearing(firstReferenceLocation, firstReferencePoint, secondReferenceLocation, secondReferencePoint);
    assertEquals(353.84, bearing, 0.001);
    assertEquals(bearing, beaconMapBackground.getBearing(), 0.001);
}
```

Other implementations ([1](https://gist.github.com/conorbuck/2606166) and [2](https://stackoverflow.com/questions/46590154/calculate-bearing-between-2-points-with-javascript)) look equal to the current one.

File: app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMapBackground.java (hunk 4 above)
Comment (Steppschuh) [reply app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMapBackground.java]: I guess that relates to the starting point of the circle (0°) being on the x axis (east, so 90° shifted from north), but we use bearing, where 0° should be north. It's also used in `getBearing`:

```java
public static double getBearing(@NonNull Point firstPoint, @NonNull Point secondPoint) {
    double angle = Math.atan2(secondPoint.y - firstPoint.y, secondPoint.x - firstPoint.x) * 180 / Math.PI;
    return (angle + 90) % 360;
}
```

File: app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMapBackground.java (hunk 1 above)
Comment (Steppschuh) [reply app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/map/BeaconMapBackground.java]: Fixed

**Pull request merged**